A29-xds-tls-security.md

```diff
@@ -0,0 +1,543 @@
+to send the certificate.
+
+As part of validating all CDS updates, if any of the following fields are present, gRPC
+will NACK a CDS update:
+
+* [sni][SNI]
+* [allow_renegotiation][ALLOW-RENEG]
+* [max_session_keys][MAX-SESS-KEYS]
+
+[A27:CDS]: A27-xds-global-load-balancing.md#cds
+[CL-TS]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L937
+[CL-TS-comment]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L932
+[CL-TS-matches]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L680
+[CTC]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L37
+[SNI]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L40
+[ALLOW-RENEG]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L47
+[MAX-SESS-KEYS]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L53
+
+##### Server Authorization aka Subject-Alt-Name Checks
+
+"Server authorization" is performed by the client where the client authorizes
+a server for the connection as described below. This check is in lieu of the
+canonical [hostname check in Web PKI][RFC6125-6].
+
+If [`match_subject_alt_names`][match_subject_alt_names] in
+[`default_validation_context`][default_validation_context] of
+[combined_validation_context][CVC] of the received `UpstreamTlsContext` is populated
+then gRPC checks the SAN entries in the leaf server certificate against the
+`match_subject_alt_names` values as follows:
+
+* if the `match_subject_alt_names` list is empty then there is nothing to match and the
+check succeeds.
+
+* if there is no leaf server certificate, or there are no SAN entries in the certificate
+the check fails.
+
+* each SAN entry of type DNS, URI and IP address is considered for the below match
+logic. An IP address is converted to its canonical string representation. For
+IPv6 this includes [maximum zero compression][zero-compr], [no leading zeros][no-leading-0]
+and [lower case][lower-case] e.g. `2001:db8::1` but *not* `2001:DB8:0::01`.
+If an entry matches as per this logic, the check completes successfully.
+
+  * if a SAN entry is empty then the match fails for that entry.
+
+  * a SAN entry is matched against each value of
+[`match_subject_alt_names`][match_subject_alt_names] as follows.
+    * A `match_subject_alt_names` value is a [`StringMatcher`][StringMatcher] and
+    the match is performed as per the semantics described for each `match_pattern`
```

Comment:
It seems DNS and exact match is supposed to do subdomain matching (e.g., `*.example.com`) as is typically done for certs.
https://github.com/envoyproxy/envoy/blob/c94e646e0280e4c521f8e613f1ae2a02b274dbbf/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L358-L359

Envoy impl: https://github.com/envoyproxy/envoy/blob/6b575e3e353e2fcbf21f362e7c6edcc52f970846/source/extensions/transport_sockets/tls/cert_validator/default_validator.cc#L319

Reply:
Should that be treated as a bug in Envoy? When the `StringMatcher` supports SUFFIX match why should exact match perform subdomain matching?

Reply:
As discussed separately, the behavior is clearly defined in the proto and in the code. It is on purpose and allows using conventional certificates with hostname verification which may need to process wildcards. The default should be we do the same. In our discussion, it sounded like we'd handle the wildcards in this specific case as documented in the proto.

Reply:
Agreed, will modify text to match the behavior described in proto and what Envoy does

Reply:
I thought we had discussed this during implementation, and I even copied the code for doing the wildcard matching from some Java implementation that you pointed to :)

Reply:
Easwar: Correct, that's what Java had. Once we decided to support other StringMacher types (other than EXACT) I think I removed the wildcard logic from the Exact one. Anyway it looks like Go already has what Eric describes

Reply:
Just looked at Envoy and the proto and realized it is not a generalized wildcard match that is needed. There is only case that is supported and that is (as an example): `*.example.com` in the cert's DNS SAN name matching `api.example.com` in the matcher present in `match_subject_alt_names`. So if the cert's DNS SAN name has an in initial `*.` then it matches a subdomain present in the exact matcher provided the rest of the domain strings match. There are no other wildcard semantics supported

Reply:
Why do you say that it is not a generalized wildcard match?

```
  // When a certificate has wildcard DNS SAN entries, to match a specific client, it should be
  // configured with exact match type in the :ref:`string matcher <envoy_v3_api_msg_type.matcher.v3.StringMatcher>`.
  // For example if the certificate has "\*.example.com" as DNS SAN entry, to allow only "api.example.com",
  // it should be configured as shown below.
  //
  // .. code-block:: yaml
  //
  //  match_subject_alt_names:
  //    exact: "api.example.com"
```

This is just an example.

Reply:
Easwar: the "*" is only allowed in the initial position followed by "." as in "*.example.com". "*" in any other place is not processed as wildcard. So "api.*.com" is not processed as wildcard . Also "*" in the StringMatcher string has no significance for wildcard processing so the only "*." that matters is in the cert SAN string.

This is what one sees in the Envoy code

Reply:
Yes, DNS wildcard match supports a wildcard only in the first component.

Reply:
> This is what one sees in the Envoy code

I'm very concerned that generalized wildcard matching was considered. It is _not_ a string-based wildcard. This should be normal HTTP TLS matching rules. I don't think we should re-define it or reimplement it if it can be avoided. There are many edge cases like `*.example.com` does not match `sub.sub.example.com`. And some implementations appropriately disallow wildcards with TLDs like `*.com`.

https://datatracker.ietf.org/doc/html/rfc2818#page-5 talks about it some, but it is out-of-date without any updated text. https://codereview.chromium.org/762013002 and https://bugzilla.mozilla.org/show_bug.cgi?id=1107791 are the browsers removing the `f*.example.com` ability.

Reply:
Sorry for the confusion. We did *not* consider generalized wildcard matching. The implementation in Go does DNS wildcard matching. See [here](https://github.com/grpc/grpc-go/blob/master/internal/credentials/xds/handshake_info.go#L256-L305). I just got confused by Sanjay's [comment](https://github.com/grpc/proposal/pull/243#discussion_r678746033).

Reply:
Sorry for causing the confusion in the first place. And in fact not considering the proper subdomain matching spec that was evident in the proto and Envoy code.

Reply:
fixed the text